File: examples/nodejs/win32.js
const koffi = require('koffi');

const CS_VREDRAW = 0x0001;
const CS_HREDRAW = 0x0002;
const WS_OVERLAPPEDWINDOW = 0x00cf0000;
const CW_USEDEFAULT = -2147483648;
const SW_HIDE = 0;
const SW_SHOW = 5;
const PM_REMOVE = 0x0001;
const WM_DESTROY = 0x0002;
const WM_CLOSE = 0x0010;
const WM_QUIT = 0x0012;
const IDC_ARROW = 32512;
const WNDCLASSEXW_SIZE = 80;

const WndProc = koffi.proto('intptr_t __stdcall WndProc(void *hwnd, uint32_t message, uintptr_t wParam, intptr_t lParam)');

const WNDCLASSEXW = koffi.struct('WNDCLASSEXW', {
  cbSize: 'uint32_t',
  style: 'uint32_t',
  lpfnWndProc: koffi.pointer(WndProc),
  cbClsExtra: 'int32_t',
  cbWndExtra: 'int32_t',
  hInstance: 'void *',
  hIcon: 'void *',
  hCursor: 'void *',
  hbrBackground: 'void *',
  lpszMenuName: 'const char16_t *',
  lpszClassName: 'const char16_t *',
  hIconSm: 'void *'
});

const POINT = koffi.struct('POINT', {
  x: 'int32_t',
  y: 'int32_t'
});

const MSG = koffi.struct('MSG', {
  hwnd: 'void *',
  message: 'uint32_t',
  wParam: 'uintptr_t',
  lParam: 'intptr_t',
  time: 'uint32_t',
  pt: POINT,
  lPrivate: 'uint32_t'
});

const kernel32 = koffi.load('kernel32.dll');
const user32 = koffi.load('user32.dll');

const GetModuleHandleW = kernel32.func('void * __stdcall GetModuleHandleW(const char16_t *lpModuleName)');
const LoadCursorW = user32.func('void * __stdcall LoadCursorW(void *hInstance, intptr_t lpCursorName)');
const RegisterClassExW = user32.func('uint16_t __stdcall RegisterClassExW(const WNDCLASSEXW *wndclass)');
const CreateWindowExW = user32.func('void * __stdcall CreateWindowExW(uint32_t exStyle, const char16_t *className, const char16_t *windowName, uint32_t style, int32_t x, int32_t y, int32_t width, int32_t height, void *parent, void *menu, void *instance, void *param)');
const ShowWindow = user32.func('bool __stdcall ShowWindow(void *hwnd, int32_t cmdShow)');
const UpdateWindow = user32.func('bool __stdcall UpdateWindow(void *hwnd)');
const PeekMessageW = user32.func('bool __stdcall PeekMessageW(_Out_ MSG *lpMsg, void *hwnd, uint32_t minFilter, uint32_t maxFilter, uint32_t removeMsg)');
const TranslateMessage = user32.func('bool __stdcall TranslateMessage(const MSG *lpMsg)');
const DispatchMessageW = user32.func('intptr_t __stdcall DispatchMessageW(const MSG *lpMsg)');
const DefWindowProcW = user32.func('intptr_t __stdcall DefWindowProcW(void *hwnd, uint32_t msg, uintptr_t wParam, intptr_t lParam)');
const DestroyWindow = user32.func('bool __stdcall DestroyWindow(void *hwnd)');
const PostQuitMessage = user32.func('void __stdcall PostQuitMessage(int32_t exitCode)');
const UnregisterClassW = user32.func('bool __stdcall UnregisterClassW(const char16_t *className, void *instance)');

class Win32Window {
  constructor(title, width, height, visible, classPrefix) {
    this.title = title;
    this.width = width;
    this.height = height;
    this.visible = visible;
    this.className = `${classPrefix}_${process.pid}`;
    this.instance = GetModuleHandleW(null);
    this.hwnd = null;
    this.hwndValue = 0n;
    this.closed = false;
    this.classRegistered = false;

    this.wndProcPointer = koffi.register((hwnd, message, wParam, lParam) => this.#windowProc(hwnd, message, wParam, lParam), koffi.pointer(WndProc));
  }

  create() {
    const cursor = LoadCursorW(null, IDC_ARROW);
    const klass = {
      cbSize: WNDCLASSEXW_SIZE,
      style: CS_HREDRAW | CS_VREDRAW,
      lpfnWndProc: this.wndProcPointer,
      cbClsExtra: 0,
      cbWndExtra: 0,
      hInstance: this.instance,
      hIcon: null,
      hCursor: cursor,
      hbrBackground: null,
      lpszMenuName: null,
      lpszClassName: this.className,
      hIconSm: null
    };

    const atom = RegisterClassExW(klass);
    if (atom === 0) {
      throw new Error('RegisterClassExW failed.');
    }
    this.classRegistered = true;

    const hwnd = CreateWindowExW(
      0,
      this.className,
      this.title,
      WS_OVERLAPPEDWINDOW,
      CW_USEDEFAULT,
      CW_USEDEFAULT,
      this.width,
      this.height,
      null,
      null,
      this.instance,
      null
    );

    if (!hwnd) {
      this.destroy();
      throw new Error('CreateWindowExW failed.');
    }

    this.hwnd = hwnd;
    this.hwndValue = koffi.address(hwnd);
    ShowWindow(hwnd, this.visible ? SW_SHOW : SW_HIDE);
    UpdateWindow(hwnd);
    return this.hwndValue;
  }

  pumpMessages() {
    const message = {};
    while (PeekMessageW(message, null, 0, 0, PM_REMOVE)) {
      TranslateMessage(message);
      DispatchMessageW(message);
      if (message.message === WM_QUIT) {
        this.closed = true;
      }
    }
    return !this.closed;
  }

  close() {
    if (this.hwnd) {
      DestroyWindow(this.hwnd);
    }
  }

  destroy() {
    if (this.hwnd) {
      DestroyWindow(this.hwnd);
      this.hwnd = null;
      this.hwndValue = 0n;
    }
    if (this.classRegistered) {
      UnregisterClassW(this.className, this.instance);
      this.classRegistered = false;
    }
  }

  dispose() {
    this.destroy();
    if (this.wndProcPointer) {
      koffi.unregister(this.wndProcPointer);
      this.wndProcPointer = null;
    }
  }

  #windowProc(hwnd, message, wParam, lParam) {
    if (message === WM_CLOSE) {
      DestroyWindow(hwnd);
      return 0;
    }

    if (message === WM_DESTROY) {
      this.closed = true;
      if (this.hwnd && koffi.address(this.hwnd) === koffi.address(hwnd)) {
        this.hwnd = null;
        this.hwndValue = 0n;
      }
      PostQuitMessage(0);
      return 0;
    }

    return DefWindowProcW(hwnd, message, wParam, lParam);
  }
}

module.exports = {
  Win32Window
};

File: examples/rust/src/win32.rs
#![cfg(windows)]

use std::error::Error;
use std::ffi::c_void;
use std::ptr::null_mut;

use windows_sys::Win32::Foundation::{GetLastError, HINSTANCE, HWND, LPARAM, LRESULT, WPARAM};
use windows_sys::Win32::System::LibraryLoader::GetModuleHandleW;
use windows_sys::Win32::UI::WindowsAndMessaging::{
    CreateWindowExW, DefWindowProcW, DestroyWindow, DispatchMessageW, GetWindowLongPtrW,
    LoadCursorW, PeekMessageW, PostQuitMessage, RegisterClassExW, SetWindowLongPtrW, ShowWindow,
    TranslateMessage, UnregisterClassW, CREATESTRUCTW, CW_USEDEFAULT, GWLP_USERDATA, IDC_ARROW,
    MSG, PM_REMOVE, SW_HIDE, SW_SHOW, WM_CLOSE, WM_DESTROY, WM_NCCREATE, WM_QUIT, WNDCLASSEXW,
    WS_OVERLAPPEDWINDOW,
};

pub type AnyResult<T> = Result<T, Box<dyn Error>>;

pub struct Win32Window {
    title: String,
    width: i32,
    height: i32,
    visible: bool,
    instance: HINSTANCE,
    class_name: Vec<u16>,
    title_wide: Vec<u16>,
    hwnd: HWND,
    closed: bool,
    class_registered: bool,
}

impl Win32Window {
    pub fn new(
        title: &str,
        width: i32,
        height: i32,
        visible: bool,
        class_prefix: &str,
    ) -> AnyResult<Self> {
        let instance = unsafe { GetModuleHandleW(std::ptr::null()) };
        if instance.is_null() {
            return Err(format!("GetModuleHandleW failed: {}", unsafe { GetLastError() }).into());
        }

        let class_name = to_wide(&format!("{}_{}", class_prefix, std::process::id()));
        let title_wide = to_wide(title);

        Ok(Self {
            title: title.to_string(),
            width,
            height,
            visible,
            instance,
            class_name,
            title_wide,
            hwnd: null_mut(),
            closed: false,
            class_registered: false,
        })
    }

    pub fn create(&mut self) -> AnyResult<HWND> {
        let cursor = unsafe { LoadCursorW(null_mut(), IDC_ARROW) };
        let class = WNDCLASSEXW {
            cbSize: std::mem::size_of::<WNDCLASSEXW>() as u32,
            style: 0x0001 | 0x0002,
            lpfnWndProc: Some(window_proc),
            cbClsExtra: 0,
            cbWndExtra: 0,
            hInstance: self.instance,
            hIcon: null_mut(),
            hCursor: cursor,
            hbrBackground: null_mut(),
            lpszMenuName: std::ptr::null(),
            lpszClassName: self.class_name.as_ptr(),
            hIconSm: null_mut(),
        };

        let atom = unsafe { RegisterClassExW(&class) };
        if atom == 0 {
            return Err(format!("RegisterClassExW failed: {}", unsafe { GetLastError() }).into());
        }
        self.class_registered = true;

        let hwnd = unsafe {
            CreateWindowExW(
                0,
                self.class_name.as_ptr(),
                self.title_wide.as_ptr(),
                WS_OVERLAPPEDWINDOW,
                CW_USEDEFAULT,
                CW_USEDEFAULT,
                self.width,
                self.height,
                null_mut(),
                null_mut(),
                self.instance,
                self as *mut Self as *mut c_void,
            )
        };

        if hwnd.is_null() {
            self.destroy();
            return Err(format!("CreateWindowExW failed: {}", unsafe { GetLastError() }).into());
        }

        self.hwnd = hwnd;
        unsafe {
            ShowWindow(hwnd, if self.visible { SW_SHOW } else { SW_HIDE });
        }
        Ok(hwnd)
    }

    pub fn handle(&self) -> HWND {
        self.hwnd
    }

    pub fn title(&self) -> &str {
        &self.title
    }

    pub fn pump_messages(&mut self) -> bool {
        let mut message: MSG = unsafe { std::mem::zeroed() };
        loop {
            let has_message = unsafe { PeekMessageW(&mut message, null_mut(), 0, 0, PM_REMOVE) };
            if has_message == 0 {
                break;
            }
            unsafe {
                TranslateMessage(&message);
                DispatchMessageW(&message);
            }
            if message.message == WM_QUIT {
                self.closed = true;
            }
        }
        !self.closed
    }

    pub fn close(&mut self) {
        if !self.hwnd.is_null() {
            unsafe {
                DestroyWindow(self.hwnd);
            }
            self.hwnd = null_mut();
        }
    }

    pub fn destroy(&mut self) {
        if !self.hwnd.is_null() {
            unsafe {
                DestroyWindow(self.hwnd);
            }
            self.hwnd = null_mut();
        }
        if self.class_registered {
            unsafe {
                UnregisterClassW(self.class_name.as_ptr(), self.instance);
            }
            self.class_registered = false;
        }
    }
}

impl Drop for Win32Window {
    fn drop(&mut self) {
        self.destroy();
    }
}

unsafe extern "system" fn window_proc(
    hwnd: HWND,
    message: u32,
    w_param: WPARAM,
    l_param: LPARAM,
) -> LRESULT {
    if message == WM_NCCREATE {
        let create_struct = l_param as *const CREATESTRUCTW;
        if !create_struct.is_null() {
            let window = (*create_struct).lpCreateParams as *mut Win32Window;
            SetWindowLongPtrW(hwnd, GWLP_USERDATA, window as isize);
        }
    }

    let window_ptr = GetWindowLongPtrW(hwnd, GWLP_USERDATA) as *mut Win32Window;
    if !window_ptr.is_null() {
        let window = &mut *window_ptr;
        match message {
            WM_CLOSE => {
                DestroyWindow(hwnd);
                return 0;
            }
            WM_DESTROY => {
                window.closed = true;
                if window.hwnd == hwnd {
                    window.hwnd = null_mut();
                }
                PostQuitMessage(0);
                return 0;
            }
            _ => {}
        }
    }

    DefWindowProcW(hwnd, message, w_param, l_param)
}

fn to_wide(value: &str) -> Vec<u16> {
    value.encode_utf16().chain(std::iter::once(0)).collect()
}
